### PR TITLE
Return a copy of messages in MockMTAppender

### LINKED ISSUE #94559: [CI] InternalEngineTests testMergeThreadLogging failing
Looks like an unsynchronized access to a synchronized list (`MockMTAppender.messages`, calling the `stream` method resulting in an unsync'd spliteration)

**Build scan:**
https://gradle-enterprise.elastic.co/s/jtsvysapo6oco/tests/:server:test/org.elasticsearch.index.engine.InternalEngineTests/testMergeThreadLogging

**Reproduction line:**
```
./gradlew ':server:test' --tests "org.elasticsearch.index.engine.InternalEngineTests.testMergeThreadLogging" -Dtests.seed=954AA314D19CFC47 -Dtests.locale=ar-QA -Dtests.timezone=America/Rainy_River -Druntime.java=19
```

**Applicable branches:**
main

**Reproduces locally?:**
Didn't try

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.index.engine.InternalEngineTests&tests.test=testMergeThreadLogging

**Failure excerpt:**
```
java.util.ConcurrentModificationException: (No message provided)

  at __randomizedtesting.SeedInfo.seed([954AA314D19CFC47:5017D13CA95EED56]:0)
  at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1631)
  at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
  at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
  at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:575)
  at java.util.stream.AbstractPipeline.evaluateToArrayNode(AbstractPipeline.java:260)
  at java.util.stream.ReferencePipeline.toArray(ReferencePipeline.java:616)
  at java.util.stream.ReferencePipeline.toArray(ReferencePipeline.java:622)
  at java.util.stream.ReferencePipeline.toList(ReferencePipeline.java:627)
  at org.elasticsearch.index.engine.InternalEngineTests.lambda$testMergeThreadLogging$39(InternalEngineTests.java:2595)
  at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:1140)
  at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:1125)
  at org.elasticsearch.index.engine.InternalEngineTests.testMergeThreadLogging(InternalEngineTests.java:2594)
  at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
  at java.lang.reflect.Method.invoke(Method.java:578)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:48)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
  at java.lang.Thread.run(Thread.java:1589)

```

### PR DESCRIPTION
**Pull request by pxsalehi**
Couldn't reproduce the failure, but in principle there is an issue there. Other options are a CopyOnWriteArrayList or just synchronizing on `messages()` return value. Returning a copy should be enough, since this is only used in one place for asserting some test expectations in an `assertBusy`.

Closes https://github.com/elastic/elasticsearch/issues/94559

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java [hunk 1]
```diff
@@ -2541,7 +2541,7 @@ private static class MockMTAppender extends AbstractAppender {
         private final List<String> messages = Collections.synchronizedList(new ArrayList<>());
 
         List<String> messages() {
-            return messages;
+            return new ArrayList<>(messages);
```
Comment (idegtiarenko) [inline server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java:2544]: NIT: should we do `List.copyOf(..)` to have it immutable?

**Review by idegtiarenko — APPROVED** [review]

**Comment by pxsalehi** [comment]
Thanks, Ievgen!

**Comment by pxsalehi** [comment]
unrelated failure. opened https://github.com/elastic/elasticsearch/issues/94987

**Comment by pxsalehi** [comment]
@elasticmachine update branch

**Pull request merged**